springboot/src/main/java/ca/jrvs/apps/trading/dao/MarketDataDao.java

```diff
@@ -0,0 +1,233 @@
+        IEX_BATCH_URL = marketDataConfig.getHost() + IEX_BATCH_PATH + marketDataConfig.getToken();
+    }
+
+    public static void main(String[] args) {
+        PoolingHttpClientConnectionManager cm = new PoolingHttpClientConnectionManager();
+        cm.setMaxTotal(50);
+        cm.setDefaultMaxPerRoute(50);
+        MarketDataConfig marketDataConfig = new MarketDataConfig();
+        marketDataConfig.setHost("https://cloud.iexapis.com/v1");
+        marketDataConfig.setToken(System.getenv("IEX_PUB_TOKEN"));
+        MarketDataDao dao = new MarketDataDao(cm, marketDataConfig);
+        String ticker = "FB";
+        IexQuote iexQuote = dao.findById(ticker).get();
+        //System.out.println(iexQuote.getSymbol());
+        List<IexQuote> quoteList = dao.findAllById(Arrays.asList("AAPL", "FB"));
+        for (int i = 0; i < quoteList.size(); i++) {
+            System.out.println(quoteList.get(i).getSymbol());
+        }
+    }
+
+    /**
+     * Borrow a HTTP client from the httpClientConnectionManger
+     *
+     * @return a httpClient
+     */
+    private CloseableHttpClient getHttpClient() {
+        return HttpClients.custom().setConnectionManager(httpClientConnectionManager)
+                .setConnectionManagerShared(true).build();
+        //prevent connectionManager shutdown when calling httpClient.close()
+    }
+
+    /**
+     * Execute a get and return http entity/body as a string
+     * <p>
+     * Use EntityUtils.toString to process HTTP entity
+     *
+     * @param url resource URL
+     * @return http response body or Optional.empty for 404 response
+     * @throws DataRetrievalFailureException if HTTP failed or status code if unexpected
+     */
+    private Optional<String> executeHttpGet(String url) throws IOException, URISyntaxException {
+        HttpClient httpClient = getHttpClient();
+        HttpGet request = new HttpGet(url);
+        HttpResponse response;
+        try {
+            response = httpClient.execute(request);
+        } catch (IOException e) {
+            throw new MalformedURLException("Invalid URL" + e.getMessage());
+        }
+        int status = response.getStatusLine().getStatusCode();
+
+        Optional<String> jsonStr;
+
+        if (status != 200) {
+            try {
+                System.out.println(EntityUtils.toString(response.getEntity()));
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to convert entity to String", e);
+            }
+            throw new DataRetrievalFailureException("Unexpected HTTP status " + status);
+        }
+        if (response.getEntity() == null) {
+            throw new RuntimeException("Empty response body");
+        }
+
+        try {
+            jsonStr = Optional.of(EntityUtils.toString(response.getEntity()));
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to convert entity to String", e);
+        }
+        return jsonStr;
```

Comment:
throwing runtime exception is a bad practice.
use DataRetrievalFailureException instead

Reply:
```
  /**
   * Execute a get and return http entity/body as a string
   *
   * Tip: use EntityUtils.toString to process HTTP entity
   *
   * @param url resource URL
   * @return http response boy or Optional.empty for 404 response
   * @throws DataRetrievalFailureException if HTTP failed or status code is unexpected
   */
  private Optional<String> executeHttpGet(String url) {
    try (CloseableHttpClient httpClient = getHttpClient()) {
      HttpGet httpGet = new HttpGet(url);
      try (CloseableHttpResponse response = httpClient.execute(httpGet)) {
        switch (response.getStatusLine().getStatusCode()) {
          case 200:
            //EntityUtils toString will also close inputStream in Entity
            String body = EntityUtils.toString(response.getEntity());
            return Optional.ofNullable(body);
          case 404:
            return Optional.empty();
          default:
            throw new DataRetrievalFailureException(
                "Unexpected status:" + response.getStatusLine().getStatusCode());
        }
      }
    } catch (IOException e) {
      throw new DataRetrievalFailureException("Unable Http execution error", e);
    }
  }
```